infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationManager.java

```diff
@@ -0,0 +1,300 @@
+            log.warn("Runtime not found to remote cluster {}", remoteClusterId);
+        }
+    }
+
+    /**
+     * The notification of change of adding/removing standby's without epoch change.
+     *
+     * @param newConfig has the same topologyConfigId as the current config
+     */
+    public void processStandbyChange(TopologyDescriptor newConfig) {
+        if (newConfig.getTopologyConfigId() != topologyDescriptor.getTopologyConfigId()) {
+            log.error("Detected changes in the topology. The new topology descriptor {} doesn't have the same " +
+                    "siteConfigId as the current one {}", newConfig, topologyDescriptor);
+            return;
+        }
+
+        Map<String, ClusterDescriptor> newStandbys = newConfig.getStandbyClusters();
+        Map<String, ClusterDescriptor> currentStandbys = topologyDescriptor.getStandbyClusters();
+        newStandbys.keySet().retainAll(currentStandbys.keySet());
+        Set<String> standbysToRemove = currentStandbys.keySet();
+        standbysToRemove.removeAll(newStandbys.keySet());
+
+        /*
+         * Remove standbys that are not in the new config
+         */
+        for (String siteID : standbysToRemove) {
+            stopLogReplicationRuntime(siteID);
+            topologyDescriptor.removeStandbySite(siteID);
+        }
+
+        //Start the standbys that are in the new config but not in the current config
+        for (String clusterId : newConfig.getStandbyClusters().keySet()) {
+            if (runtimeToRemoteCluster.get(clusterId) == null) {
+                ClusterDescriptor clusterInfo = newConfig.getStandbyClusters().get(clusterId);
+                topologyDescriptor.addStandbySite(clusterInfo);
+                startLogReplicationRuntime(clusterInfo);
+            }
+        }
+    }
+
+    /**
+     * Query max stream tail for all streams to be replicated.
+     *
+     * @return max tail of all relevant streams.
+     */
+    private long queryStreamTail() {
+        Set<String> streamsToReplicate = logReplicationConfig.getStreamsToReplicate();
+        long maxTail = Address.NON_ADDRESS;
+        for (String s : streamsToReplicate) {
+            UUID currentUUID = CorfuRuntime.getStreamID(s);
+            Map<UUID, Long> tailMap = corfuRuntime.getAddressSpaceView().getAllTails().getStreamTails();
+            Long currentTail = tailMap.get(currentUUID);
+            if (currentTail != null) {
+                maxTail = Math.max(maxTail, currentTail);
+            }
+        }
+        return maxTail;
+    }
+
+    /**
+     * Given a timestamp, calculate how many entries to be sent for all replicated streams.
+     *
+     * @param timestamp
+     */
+    private long queryEntriesToSend(long timestamp) {
+        int totalNumEntries = 0;
+
+        for (CorfuLogReplicationRuntime runtime: runtimeToRemoteCluster.values()) {
+            totalNumEntries += runtime.getNumEntriesToSend(timestamp);
+        }
+
+        return totalNumEntries;
+    }
```

Comment:
I think this numEntries is not exactly how many entries we need to send. Because we only transfer some streams, and `getNumEntriesToSend` returns (maxStreamTail - ackedTimeStamp), which will count other log entries. Besides, for loop will amplify totalNumEntries multiple times.

Reply:
For example, the timestamp is 20, and we are interested in stream A[1, 3, 4], B[2, 5, 6], C[7, 8, 20]
The result will be (20-4) + (20-6) + (20-20) = 30. The expected result will be 0.

Reply:
+ @xiaoqin2012 please address this

Reply:
Adding a comment in the code so @zhangn49 and @xiaoqin2012 can work on this!